### PR TITLE
🐛 Add ability for nodeToMachines to filter by clusterName and namespace

### LINKED ISSUE #3831: Node to Machine mapping function should filter by namespace and cluster
**Detailed Description**

This came out from https://github.com/kubernetes-sigs/cluster-api/pull/3826#discussion_r508557322, we need to audit and see where we use a mapping function from node -> machine and filter out on cluster and namespace.

- For cluster filtering, we can achieve this today by using a label selector.
- For namespace filtering, we need instead to wait for controller runtime to support this functionality.

/kind cleanup
/milestone v0.4.0
/cc @ncdc

### PR DESCRIPTION
**Pull request by enxebre**
**What this PR does / why we need it**:
This is a best effort to filter machines by clusterName and namespace when reconciling from a node event.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3831

### CODE REVIEW

**Comment by mrajashree** [comment]
lgtm

File: controllers/machine_controller.go [hunk 1]
```diff
@@ -657,12 +657,32 @@ func (r *MachineReconciler) nodeToMachine(o client.Object) []reconcile.Request {
...
+
+	// Match by clusterName when the node has the annotation.
+	clusterName, ok := node.GetAnnotations()[clusterv1.ClusterNameAnnotation]
+	if ok {
+		filters = append(filters,
+			client.MatchingLabels{
+				clusterv1.ClusterLabelName: clusterName,
+			},
+		)
+	}
```
Comment (vincepri) [inline controllers/machine_controller.go:673]: Add a test?

File: controllers/machine_controller.go [hunk 2]
```diff
@@ -657,12 +657,32 @@ func (r *MachineReconciler) nodeToMachine(o client.Object) []reconcile.Request {
...
+
+	// Match by namespace when the node has the annotation.
+	namespace, ok := node.GetAnnotations()[clusterv1.ClusterNamespaceAnnotation]
+	if ok {
```
Comment (vincepri) [inline controllers/machine_controller.go:677]:
```suggestion
	if namespace, ok := node.GetAnnotations()[clusterv1.ClusterNamespaceAnnotation]; ok {
```

Comment (vincepri) [reply controllers/machine_controller.go:677]: Add a test

File: controllers/machine_controller.go (hunk 1 above)
Comment (vincepri) [reply controllers/machine_controller.go:673]: These annotations were added later, so I'd like to make sure we cover:

- A node with none of these annotations
- A node with a single annotation
- A node with both annotations

**Review by vincepri — CHANGES_REQUESTED** [review]

**Comment by enxebre** [comment]
@vincepri added tests for all the cases. I used a testEnv since fakeClient does not support indexes.
I also created https://github.com/kubernetes-sigs/cluster-api/issues/4518 and so then we can also filter by providerID.

File: controllers/machine_controller_test.go [hunk 3]
```diff
@@ -1626,6 +1626,270 @@ func TestIsDeleteNodeAllowed(t *testing.T) {
...
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "machine-reconcile-",
+			Namespace:    ns.Name,
+		},
+	}
+
+	targetNode := &corev1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "node-1",
+			Namespace: ns.Name,
```
Comment (sbueringer) [inline controllers/machine_controller_test.go:1704]: As Nodes are cluster-wide setting the Namespace doesn't make a lot of sense imho. Should we also prefix this again to make sure we don't clash with other tests?

File: controllers/machine_controller_test.go [hunk 4]
```diff
@@ -1626,6 +1626,270 @@ func TestIsDeleteNodeAllowed(t *testing.T) {
...
+	key = client.ObjectKey{Name: randomMachine.Name, Namespace: randomMachine.Namespace}
+	g.Eventually(func() bool {
+		if err := testEnv.Get(ctx, key, randomMachine); err != nil {
+			return false
+		}
+		return randomMachine.Status.NodeRef != nil
+	}, timeout).Should(BeTrue())
+
+	// Fake nodes for actual test of nodeToMachine.
+	FakeNodes := []*corev1.Node{
```
Comment (sbueringer) [inline controllers/machine_controller_test.go:1831]:
```suggestion
	fakeNodes := []*corev1.Node{
```

**Review by vincepri — APPROVED** [review]
/approve

**Pull request merged**